<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Workflow · JWAS.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="JWAS.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">JWAS.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../theory/theory/">Some Theory</a></li><li><a class="tocitem" href="../../citing/citing/">Citing</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../getstarted/">Get Started</a></li><li class="is-active"><a class="tocitem" href>Workflow</a><ul class="internal"><li><a class="tocitem" href="#Available-Models"><span>Available Models</span></a></li><li><a class="tocitem" href="#Get-Data-Ready"><span>Get Data Ready</span></a></li><li><a class="tocitem" href="#Step-1:-Load-Packages"><span>Step 1: Load Packages</span></a></li><li><a class="tocitem" href="#Step-2:-Read-Data"><span>Step 2: Read Data</span></a></li><li><a class="tocitem" href="#Step-3:-Build-Model-Equations"><span>Step 3: Build Model Equations</span></a></li><li><a class="tocitem" href="#Step-4:-Set-Factors-or-Covariate"><span>Step 4: Set Factors or Covariate</span></a></li><li><a class="tocitem" href="#Step-5:-Set-Random-or-Fixed-Effects"><span>Step 5: Set Random or Fixed Effects</span></a></li><li><a class="tocitem" href="#Step-6:-Run-Bayesian-Analysis"><span>Step 6: Run Bayesian Analysis</span></a></li><li><a class="tocitem" href="#check-results"><span>check results</span></a></li></ul></li><li><a class="tocitem" href="../public/">Public</a></li><li><a class="tocitem" href="../internals/">Internals</a></li></ul></li><li><span class="tocitem">Mixed Effects Neural Networks (NNMM)</span><ul><li><a class="tocitem" href="../../nnmm/Part1_introduction/">Part 1. introduction</a></li><li><a class="tocitem" href="../../nnmm/Part2_NNMM/">Part 2. NNMM</a></li><li><a class="tocitem" href="../../nnmm/Part3_NNMMwithIntermediateOmics/">Part 3. NNMM with intermediate omics</a></li><li><a class="tocitem" href="../../nnmm/Part4_PartialConnectedNeuralNetwork/">Part 4. partial connected neural network</a></li><li><a class="tocitem" href="../../nnmm/Part5_UserDefinedNonlinearFunction/">Part 5. user-defined onlinear function</a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/examples/">Examples</a></li></ul></li><li><a class="tocitem" href="../../FrequentlyAskedQuestions/FrequentlyAskedQuestions/">Frequently Asked Questions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Workflow</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Workflow</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/reworkhow/JWAS.jl/blob/master/docs/src/manual/workflow.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Workflow"><a class="docs-heading-anchor" href="#Workflow">Workflow</a><a id="Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow" title="Permalink"></a></h1><p>A step by step workflow for how to run JWAS is shown in this section. The workflow below is used to demonstrate a three-trait Bayesian linear mixed model fitting fixed effects (x1, x3), random effects (x2), direct genetic effects (ID), maternal genetic effects (dam) and genomic information.</p><ul><li><a href="#Workflow">Workflow</a></li><ul><li><a href="#Available-Models">Available Models</a></li><li><a href="#Get-Data-Ready">Get Data Ready</a></li><li><a href="#Step-1:-Load-Packages">Step 1: Load Packages</a></li><li><a href="#Step-2:-Read-Data">Step 2: Read Data</a></li><li><a href="#Step-3:-Build-Model-Equations">Step 3: Build Model Equations</a></li><li><a href="#Step-4:-Set-Factors-or-Covariate">Step 4: Set Factors or Covariate</a></li><li><a href="#Step-5:-Set-Random-or-Fixed-Effects">Step 5: Set Random or Fixed Effects</a></li><li><a href="#Step-6:-Run-Bayesian-Analysis">Step 6: Run Bayesian Analysis</a></li><li><a href="#check-results">check results</a></li></ul></ul><h2 id="Available-Models"><a class="docs-heading-anchor" href="#Available-Models">Available Models</a><a id="Available-Models-1"></a><a class="docs-heading-anchor-permalink" href="#Available-Models" title="Permalink"></a></h2><p>Given the data and model equations, several different types of models are available in JWAS as shown below. In the table below, &quot;X&quot; denotes the type of available data, and &quot;Y&lt;=A&quot; denotes that Y individuals is a subset of A individuals.  </p><table><tr><th style="text-align: center">Linear Mixed Models (LMM)</th><th style="text-align: center">phenotypes (Y)</th><th style="text-align: center">pedigree (A)</th><th style="text-align: center">genotypes (G)</th><th style="text-align: center">notes</th></tr><tr><td style="text-align: center">Conventional LMM</td><td style="text-align: center">X</td><td style="text-align: center"></td><td style="text-align: center"></td><td style="text-align: center"></td></tr><tr><td style="text-align: center">Pedigree-based LMM</td><td style="text-align: center">X</td><td style="text-align: center">X</td><td style="text-align: center"></td><td style="text-align: center">Y&lt;=A</td></tr><tr><td style="text-align: center">Complete Genomic LMM</td><td style="text-align: center">X</td><td style="text-align: center">optional</td><td style="text-align: center">X</td><td style="text-align: center">Y&lt;=G</td></tr><tr><td style="text-align: center">Incomplete Genomic LMM</td><td style="text-align: center">X</td><td style="text-align: center">X</td><td style="text-align: center">X</td><td style="text-align: center">Y&lt;=A,G&lt;=A</td></tr></table><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><ul><li><p><strong>Incomplete Genomic LMM</strong> is also called &quot;single-step&quot; methods in animal breeding.</p></li><li><p>Pedigree information may be used in <strong>Complete Genomic LMM</strong> for extra polygenic effects to account for genetic variance not explained by the genomic data (e.g., SNPs).</p></li><li><p><strong>Pedigree-based LMM</strong> (none of the individuals in the pedigree are genotyped) and <strong>Complete Genomic LMM</strong> (all individuals in the pedigree are genotyped) are special cases of <strong>Incomplete Genomic LMM</strong> (part of the individuals in the pedigree are genotyped).</p></li></ul></div></div><h2 id="Get-Data-Ready"><a class="docs-heading-anchor" href="#Get-Data-Ready">Get Data Ready</a><a id="Get-Data-Ready-1"></a><a class="docs-heading-anchor-permalink" href="#Get-Data-Ready" title="Permalink"></a></h2><p>By default, input data files are comma-separated values (CSV) files, where each line of the file consists of one or more fields, separated by <strong>commas</strong>. Other field separators such as space (&#39; &#39;) or tab (&#39;\t&#39;) can be used if you supply the keyword argument, e.g, <code>CSV.read(...,delim=&#39;\t&#39;)</code> or <code>add_genotypes(...,separator=&#39;\t&#39;)</code></p><p>Click on the buttons inside the tabbed menu to see the data:</p><head>
<meta name="viewport" content="width=device-width, initial-scale=1">
<style>
body {font-family: Arial;}

/* Style the tab */
.tab {
    overflow: hidden;
    border: 1px solid #ccc;
    background-color: #f1f1f1;
}

/* Style the buttons inside the tab */
.tab button {
    background-color: inherit;
    float: left;
    border: none;
    outline: none;
    cursor: pointer;
    padding: 14px 16px;
    transition: 0.3s;
    font-size: 17px;
}

/* Change background color of buttons on hover */
.tab button:hover {
    background-color: #ddd;
}

/* Create an active/current tablink class */
.tab button.active {
    background-color: #ccc;
}

/* Style the tab content */
.tabcontent {
    display: none;
    padding: 6px 12px;
    border: 1px solid #ccc;
    border-top: none;
}
</style>
</head>
<body>

<div class="tab">
  <button class="tablinks" onclick="openCity(event, 'phenotypes')">phenotypes.txt</button>
  <button class="tablinks" onclick="openCity(event, 'pedigree')">pedigree.txt</button>
  <button class="tablinks" onclick="openCity(event, 'genotypes')">genotypes.txt</button>
  <button class="tablinks" onclick="openCity(event, 'map file')">map.txt</button>
</div>

<div id="phenotypes" class="tabcontent">
<p>ID,y1,y2,y3,x1,x2,x3,dam</p>
<p>a1,-0.06,3.58,-1.18,0.9,2,m,NA</p>
<p>a2,-0.6,4.9,0.88,0.3,1,f,NA</p>
<p>a3,-2.07,3.19,0.73,0.7,2,f,NA</p>
<p>a4,-2.63,6.97,-0.83,0.6,1,m,a2</p>
<p>a5,2.31,3.5,-1.52,0.4,2,m,a2</p>
<p>a6,0.93,4.87,-0.01,05,2,f,a3</p>
<p>a7,-0.69,3.1,-1.47,0.5,2,f,a3</p>
<p>a8,-4.69,7.31,-1.09,0.3,2,m,a6</p>
<p>a9,-2.81,7.18,0.76,0.4,2,m,a6</p>
<p>a10,1.92,1.78,-0.88,0.2,1,m,a7</p>
</div>

<div id="pedigree" class="tabcontent">
<p>ID,Sire,Dam</p>
<p>a1,0,0</p>
<p>a2,0,0</p>
<p>a3,0,0</p>
<p>a4,a1,a2</p>
<p>a5,a1,a2</p>
<p>a6,a1,a3</p>
<p>a7,a1,a3</p>
<p>a8,a4,a6</p>
<p>a9,a4,a6</p>
<p>a10,a5,a7</p>
</div>

<div id="genotypes" class="tabcontent">
<p>ID,m1,m2,m3,m4,m5</p>
<p>a1,1,2,1,1,0</p>
<p>a2,2,1,1,1,1</p>
<p>a3,1,1,0,1,1</p>
<p>a4,2,2,0,1,0</p>
<p>a5,1,1,2,1,1</p>
<p>a6,2,1,0,0,0</p>
<p>a7,0,2,1,2,1</p>
<p>a8,2,2,0,0,0</p>
<p>a9,2,1,0,1,0</p>
<p>a10,0,2,2,2,1</p>
</div>

<div id="map file" class="tabcontent">
<p>markerID,chromosome,position</p>
<p>m1,1,16977</p>
<p>m2,1,434311</p>
<p>m3,1,1025513</p>
<p>m4,2,70350</p>
<p>m5,2,101135</p>
</div>


<script>
function openCity(evt, cityName) {
    var i, tabcontent, tablinks;
    tabcontent = document.getElementsByClassName("tabcontent");
    for (i = 0; i < tabcontent.length; i++) {
        tabcontent[i].style.display = "none";
    }
    tablinks = document.getElementsByClassName("tablinks");
    for (i = 0; i < tablinks.length; i++) {
        tablinks[i].className = tablinks[i].className.replace(" active", "");
    }
    document.getElementById(cityName).style.display = "block";
    evt.currentTarget.className += " active";
}
</script>
</body><h2 id="Step-1:-Load-Packages"><a class="docs-heading-anchor" href="#Step-1:-Load-Packages">Step 1: Load Packages</a><a id="Step-1:-Load-Packages-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-Load-Packages" title="Permalink"></a></h2><pre><code class="language-julia">using JWAS,CSV,DataFrames</code></pre><hr/><p>The <code>JWAS</code> package is loaded, as well as the <code>CSV</code> and <code>DataFrame</code> packages for reading text files.</p><h2 id="Step-2:-Read-Data"><a class="docs-heading-anchor" href="#Step-2:-Read-Data">Step 2: Read Data</a><a id="Step-2:-Read-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-Read-Data" title="Permalink"></a></h2><h4 id="Read-Phenotypic-Data"><a class="docs-heading-anchor" href="#Read-Phenotypic-Data">Read Phenotypic Data</a><a id="Read-Phenotypic-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Read-Phenotypic-Data" title="Permalink"></a></h4><pre><code class="language-julia">phenotypes = CSV.read(&quot;phenotypes.txt&quot;,DataFrame,delim = &#39;,&#39;,header=true,,missingstrings=[&quot;NA&quot;])
first(phenotypes)</code></pre><p>output:</p><pre><code class="language-julia">6×8 DataFrames.DataFrame
│ Row │ ID │ y1    │ y2   │ y3    │ x1  │ x2 │ x3 │ dam       │
├─────┼────┼───────┼──────┼───────┼─────┼────┼────┼───────────┤
│ 1   │ a1 │ -0.06 │ 3.58 │ -1.18 │ 0.9 │ 2  │ m  │ missing   │
│ 2   │ a2 │ -0.6  │ 4.9  │ 0.88  │ 0.3 │ 1  │ f  │ missing   │
│ 3   │ a3 │ -2.07 │ 3.19 │ 0.73  │ 0.7 │ 2  │ f  │ missing   │
│ 4   │ a4 │ -2.63 │ 6.97 │ -0.83 │ 0.6 │ 1  │ m  │ a2        │
│ 5   │ a5 │ 2.31  │ 3.5  │ -1.52 │ 0.4 │ 2  │ m  │ a2        │
│ 6   │ a6 │ 0.93  │ 4.87 │ -0.01 │ 5.0 │ 2  │ f  │ a3        │</code></pre><hr/><p>The <strong>phenotypic data</strong> is read on line 1. On line 2, the first several rows of the phenotypic data are shown.</p><h4 id="Read-Pedigree-Data"><a class="docs-heading-anchor" href="#Read-Pedigree-Data">Read Pedigree Data</a><a id="Read-Pedigree-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Read-Pedigree-Data" title="Permalink"></a></h4><pre><code class="language-julia">pedigree   = get_pedigree(&quot;pedigree.txt&quot;,separator=&quot;,&quot;,header=true)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.PedModule.get_pedigree"><code>get_pedigree</code></a></li></ul><hr/><p>The <strong>pedigree data</strong> is read on line 1.</p><h4 id="Read-Genomic-Data"><a class="docs-heading-anchor" href="#Read-Genomic-Data">Read Genomic Data</a><a id="Read-Genomic-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Read-Genomic-Data" title="Permalink"></a></h4><pre><code class="language-julia">genotypes  = get_genotypes(&quot;genotypes.txt&quot;,G,method=&quot;BayesC&quot;,separator=&quot;,&quot;,header=true) #G is optional</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.get_genotypes"><code>get_genotypes</code></a></li></ul><hr/><p>On line 1, the genomic information is read on line 1 with the genotype file.  and variance <code>G</code> (a 3x3 matrix). In Bayesian analysis, the <code>G</code> is the mean for the prior assigned for the genomic variance with degree of freedom <code>df</code>, defaulting to 4.0. If <code>G</code> is not provided, a value is calculated from responses (phenotypes).</p><h2 id="Step-3:-Build-Model-Equations"><a class="docs-heading-anchor" href="#Step-3:-Build-Model-Equations">Step 3: Build Model Equations</a><a id="Step-3:-Build-Model-Equations-1"></a><a class="docs-heading-anchor-permalink" href="#Step-3:-Build-Model-Equations" title="Permalink"></a></h2><pre><code class="language-julia">model_equation = &quot;y1 = intercept + x1 + x3 + ID + dam + genotypes
                  y2 = intercept + x1 + x2 + x3 + ID + genotypes
                  y3 = intercept + x1 + x1*x3 + x2 + ID + genotypes&quot;
model=build_model(model_equation,R) #R is optional</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.build_model"><code>build_model</code></a></li></ul><hr/><p>The model equation for a 3-trait analysis is defined on the first 3 lines.</p><ul><li>The effects fitted in the model for trait <code>y1</code> are the intercept, <code>x1</code>, <code>x3</code>, direct genetic effects (<code>ID</code>), maternal genetic effects (<code>dam</code>), and molecular marker effects (<code>genotypes</code>).</li><li>The effects fitted in the model for trait <code>y2</code> are the intercept, <code>x1</code>, <code>x2</code>, <code>x3</code>, direct genetic effects (<code>ID</code>), and molecular marker effects (<code>genotypes</code>).</li><li>The effects fitted in the model for trait <code>y3</code> are the intercept, <code>x1</code>, the interaction between <code>x1</code> and <code>x3</code>, <code>x2</code>, direct genetic effects (<code>ID</code>), and , and molecular marker effects (<code>genotypes</code>).</li></ul><p>On the last line, the model is built given the model equation and residual variance <code>R</code> (a 3x3 matrix). In Bayesian analysis, <code>R</code> is the mean for the prior assigned for the residual variance with degree of freedom <code>df</code>, defaulting to 4.0. If <code>R</code> is not provided, a value is calculated from responses (phenotypes). By default, all effects are treated as fixed and classed as factors (categorical variables) rather than covariates (quantitative variables).</p><h2 id="Step-4:-Set-Factors-or-Covariate"><a class="docs-heading-anchor" href="#Step-4:-Set-Factors-or-Covariate">Step 4: Set Factors or Covariate</a><a id="Step-4:-Set-Factors-or-Covariate-1"></a><a class="docs-heading-anchor-permalink" href="#Step-4:-Set-Factors-or-Covariate" title="Permalink"></a></h2><pre><code class="language-julia">set_covariate(model,&quot;x1&quot;)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.set_covariate"><code>set_covariate</code></a></li></ul><hr/><p>On line 1, the effect <code>x1</code> is defined to be a covariate (a quantitative variable) rather than a factor (a categorical variable).</p><h2 id="Step-5:-Set-Random-or-Fixed-Effects"><a class="docs-heading-anchor" href="#Step-5:-Set-Random-or-Fixed-Effects">Step 5: Set Random or Fixed Effects</a><a id="Step-5:-Set-Random-or-Fixed-Effects-1"></a><a class="docs-heading-anchor-permalink" href="#Step-5:-Set-Random-or-Fixed-Effects" title="Permalink"></a></h2><pre><code class="language-julia">set_random(model,&quot;x2&quot;,G1) #G1 is optional
set_random(model,&quot;ID dam&quot;,pedigree,G2) #G2 is optional</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.set_random"><code>set_random</code></a></li></ul><hr/><p>On line 1, the <code>x2</code> class effect is defined as random with variance <code>G1</code>(a 2x2 matrix). On line 2, direct genetic effects and maternal genetic effects are fitted as <code>ID</code> and <code>dam</code> with <code>G2</code> (a 4x4 matrix) and the inverse of the numerator relationship matrix defined from pedigree. In Bayesian analysis, <code>G1</code> and <code>G2</code> are the means for the priors assigned for the variances with degree of freedom <code>df</code>, defaulting to 4.0. If <code>G1</code> or <code>G2</code> is not provided, a value is calculated from responses (phenotypes).</p><h2 id="Step-6:-Run-Bayesian-Analysis"><a class="docs-heading-anchor" href="#Step-6:-Run-Bayesian-Analysis">Step 6: Run Bayesian Analysis</a><a id="Step-6:-Run-Bayesian-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Step-6:-Run-Bayesian-Analysis" title="Permalink"></a></h2><pre><code class="language-julia">outputMCMCsamples(model,&quot;dam&quot;)
out=runMCMC(model,phenotypes)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.outputMCMCsamples"><code>outputMCMCsamples</code></a></li><li>link to documentation for <a href="../public/#JWAS.runMCMC"><code>runMCMC</code></a></li></ul><hr/><p>On line 1, MCMC samples from <code>runMCMC</code> for <code>x2</code> is saved to a file, where each row represents one sample from the MCMC. On line 2, a multi-trait BayesC analysis is performed with <code>model</code> and <code>phenotypes</code> as had been defined in step 1-6. MCMC samples for marker effects, location parameters specified on line 1, and all variance components from this analysis are saved every <code>output_samples_frequency</code> iterations to files.</p><hr/><p>Several steps above can be skipped if no related information is available, e.g., step 4 is skipped if all effects are classed as factors. Several detailed examples are available in the examples section. Here is the link to documentation for all <a href="../public/#Public-functions">Public functions</a>.</p><h2 id="check-results"><a class="docs-heading-anchor" href="#check-results">check results</a><a id="check-results-1"></a><a class="docs-heading-anchor-permalink" href="#check-results" title="Permalink"></a></h2><p>Posterior means (estimate) and standard deviations (SD) of location parameters, most variance components, and marker effects are saved as the variable <code>out</code> and in text files. They can be listed and obtained as</p><pre><code class="language-julia">keys(out)

# output:
#
# Base.KeyIterator for a Dict{Any,Any} with 7 entries. Keys:
#   &quot;polygenic effects covariance matrix&quot;
#   &quot;Model frequency&quot;
#   &quot;residual covariance matrix&quot;
#   &quot;marker effects&quot;
#   &quot;marker effects variance&quot;
#   &quot;location parameters&quot;
#   &quot;Pi&quot;

out[&quot;residual variance&quot;]

# output:
#
#Covariance	Estimate	SD
#y1_y1	1.65265	0.29405
#y1_y2	-0.0290279	0.02347
#y1_y3	-0.252009	0.048289
#y2_y1	-0.0290279	0.02347
#y2_y2	0.977405	0.009732
#y2_y3	0.0451994	0.095828
#y3_y1	-0.252009	0.048289
#y3_y2	0.0451994	0.095828
#y3_y3	0.363878	0.049278
</code></pre><p>In addition, MCMC samples from posterior distributions of marker effects, all variance components, and location parameters specified in step 7, are saved to text files in your working directory. Users can compute the posterior distributions of parameters of interest using these MCMC samples files. A list of output files are shown below.</p><h3 id="Output-files:"><a class="docs-heading-anchor" href="#Output-files:">Output files:</a><a id="Output-files:-1"></a><a class="docs-heading-anchor-permalink" href="#Output-files:" title="Permalink"></a></h3><p>Below is a list of files containing estimates and standard deviations for variables of interest. </p><table><tr><th style="text-align: right">file name</th><th style="text-align: right">description</th></tr><tr><td style="text-align: right">EBV_y1.txt</td><td style="text-align: right">estimated breeding values for trait named &quot;y1&quot;</td></tr><tr><td style="text-align: right">EBV_y2.txt</td><td style="text-align: right">estimated breeding values for trait named &quot;y2&quot;</td></tr><tr><td style="text-align: right">EBV_y3.txt</td><td style="text-align: right">estimated breeding values for trait named &quot;y3&quot;</td></tr><tr><td style="text-align: right">genetic_variance.txt</td><td style="text-align: right">estimated genetic variance-covariance for all traits</td></tr><tr><td style="text-align: right">heritability.txt</td><td style="text-align: right">estimated heritability</td></tr><tr><td style="text-align: right">location_parameters.txt</td><td style="text-align: right">estimated non-genetic effects</td></tr><tr><td style="text-align: right">pi_genotypes.txt</td><td style="text-align: right">estimated pi</td></tr><tr><td style="text-align: right">polygenic<em>effects</em>covariance_matrix.txt</td><td style="text-align: right">estimated variance-covariance between polygenic effects (y1<em>ID, y2</em>ID, y3<em>ID, y1</em>dam)</td></tr><tr><td style="text-align: right">marker<em>effects</em>genotypes.txt</td><td style="text-align: right">estimated marker effects for all traits</td></tr><tr><td style="text-align: right">residual_variance.txt</td><td style="text-align: right">estimated residual variance-covariance for all traits</td></tr></table><p>Below is a list of files containing MCMC samples for variables of interest. </p><table><tr><th style="text-align: right">file name</th><th style="text-align: right">description</th></tr><tr><td style="text-align: right">MCMC<em>samples</em>EBV_y1.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of breeding values for trait named &quot;y1&quot;</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>EBV_y2.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of breeding values for trait named &quot;y2&quot;</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>EBV_y3.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of breeding values for trait named &quot;y3&quot;</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>genetic_variance.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of genetic variance-covariance for all traits</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>heritability.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of heritability</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>marker<em>effects</em>genotypes_y1</td><td style="text-align: right">MCMC samples from the posterior distribution of marker effects for trait named &quot;y1&quot;</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>marker<em>effects</em>genotypes_y2</td><td style="text-align: right">MCMC samples from the posterior distribution of marker effects for trait named &quot;y2&quot;</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>marker<em>effects</em>genotypes_y3</td><td style="text-align: right">MCMC samples from the posterior distribution of marker effects for trait named &quot;y3&quot;</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>marker<em>effects</em>variances_genotypes.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of marker effect variance for all traits</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>pi_genotypes.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of pi</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>polygenic<em>effects</em>variance.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of variance-covariance between y1<em>ID, y2</em>ID, y3<em>ID, y1</em>dam</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>residual_variance.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of residual variance-covariance for all traits</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>y1.dam.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of dam effect for y1</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>y1.ID<em>y2.ID</em>y3.ID<em>y1.dam</em>variances.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of variance-covariance between y1<em>ID, y2</em>ID, y3<em>ID, y1</em>dam</td></tr><tr><td style="text-align: right">MCMC<em>samples</em>y2.x2<em>y3.x2</em>variances.txt</td><td style="text-align: right">MCMC samples from the posterior distribution of variance-covariance between y2<em>x2 and y3</em>x2</td></tr></table></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../getstarted/">« Get Started</a><a class="docs-footer-nextpage" href="../public/">Public »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 8 March 2024 20:20">Friday 8 March 2024</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
