<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Some Theory · JWAS.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="JWAS.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">JWAS.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li class="is-active"><a class="tocitem" href>Some Theory</a><ul class="internal"><li><a class="tocitem" href="#A-Table-for-Bayesian-Linear-Mixed-Models-(BLMM)"><span>A Table for Bayesian Linear Mixed Models (BLMM)</span></a></li><li><a class="tocitem" href="#Models"><span>Models</span></a></li></ul></li><li><a class="tocitem" href="../../citing/citing/">Citing</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../manual/getstarted/">Get Started</a></li><li><a class="tocitem" href="../../manual/workflow/">Workflow</a></li><li><a class="tocitem" href="../../manual/public/">Public</a></li><li><a class="tocitem" href="../../manual/internals/">Internals</a></li></ul></li><li><span class="tocitem">Mixed Effects Neural Networks (NNMM)</span><ul><li><a class="tocitem" href="../../nnmm/Part1_introduction/">Part 1. introduction</a></li><li><a class="tocitem" href="../../nnmm/Part2_NNMM/">Part 2. NNMM</a></li><li><a class="tocitem" href="../../nnmm/Part3_NNMMwithIntermediateOmics/">Part 3. NNMM with intermediate omics</a></li><li><a class="tocitem" href="../../nnmm/Part4_PartialConnectedNeuralNetwork/">Part 4. partial connected neural network</a></li><li><a class="tocitem" href="../../nnmm/Part5_UserDefinedNonlinearFunction/">Part 5. user-defined onlinear function</a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/examples/">Examples</a></li></ul></li><li><a class="tocitem" href="../../FrequentlyAskedQuestions/FrequentlyAskedQuestions/">Frequently Asked Questions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Some Theory</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Some Theory</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/reworkhow/JWAS.jl/blob/master/docs/src/theory/theory.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Some-Theory-in-JWAS"><a class="docs-heading-anchor" href="#Some-Theory-in-JWAS">Some Theory in JWAS</a><a id="Some-Theory-in-JWAS-1"></a><a class="docs-heading-anchor-permalink" href="#Some-Theory-in-JWAS" title="Permalink"></a></h1><h2 id="A-Table-for-Bayesian-Linear-Mixed-Models-(BLMM)"><a class="docs-heading-anchor" href="#A-Table-for-Bayesian-Linear-Mixed-Models-(BLMM)">A Table for Bayesian Linear Mixed Models (BLMM)</a><a id="A-Table-for-Bayesian-Linear-Mixed-Models-(BLMM)-1"></a><a class="docs-heading-anchor-permalink" href="#A-Table-for-Bayesian-Linear-Mixed-Models-(BLMM)" title="Permalink"></a></h2><p><img src="../../assets/BLMM.png" alt="BLMM"/></p><h2 id="Models"><a class="docs-heading-anchor" href="#Models">Models</a><a id="Models-1"></a><a class="docs-heading-anchor-permalink" href="#Models" title="Permalink"></a></h2><h3 id="Complete-Genomic-Data"><a class="docs-heading-anchor" href="#Complete-Genomic-Data">Complete Genomic Data</a><a id="Complete-Genomic-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Complete-Genomic-Data" title="Permalink"></a></h3><p>The general form of the multivariate (univariate) mixed effects model for individual <span>$i$</span> from <span>$n$</span> individuals with complete genomic data in JWAS is</p><p class="math-container">\[
\mathbf{y}_{i}
 =\sum_{j=1}^{p_{b}}X_{ij}\mathbf{b}_{j}+\sum_{k=1}^{p_{u}}Z_{ik}\mathbf{u}_{k}
 +\sum_{l=1}^{p}M_{il}\boldsymbol{\alpha}_{l}+\mathbf{e}_{i}(1),\]</p><p>where <span>$\mathbf{y}_{i}$</span> is a vector of phenotypes of <span>$t$</span> traits for individual <span>$i$</span>; <span>$X_{ij}$</span> is the incidence matrix covariate corresponding to the <span>$j$</span>th fixed effect for individual <span>$i$</span>; <span>$\mathbf{b}_{j}$</span> is a vector of <span>$j$</span>th fixed effects for the <span>$t$</span> traits; <span>$Z_{ik}$</span> is the incidence matrix covariate corresponding to the <span>$k$</span>th random effect for individual <span>$i$</span>; <span>$\boldsymbol{u}_{k}$</span> is a vector of the <span>$k$</span>th random effects of <span>$t$</span> traits; <span>$M_{il}$</span> is the genotype covariate at locus <span>$l$</span> for individual <span>$i$</span>, <span>$p$</span> is the number of genotyped loci (each coded as 0,1,2), <span>$\boldsymbol{\alpha}_{l}$</span> is a vector of allele substitution effects or marker effects of <span>$t$</span> traits for locus <span>$j$</span>, and <span>$\mathbf{e}_{i}$</span> is the vector of random residual effects of <span>$t$</span> traits for individual <span>$i$</span>. The JWAS implementation of this model involves missing phenotypes being imputed at each iteration of MCMC \cite{sorensenGianolaBook} so that all individuals have observations for all traits. Note that when the number of traits <span>$t=1$</span>, the general form above simplifies to the single-trait  mixed effects model, and all vectors of effects in equation (1) become scalars.</p><h3 id="Incomplete-Genomic-Data"><a class="docs-heading-anchor" href="#Incomplete-Genomic-Data">Incomplete Genomic Data</a><a id="Incomplete-Genomic-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Incomplete-Genomic-Data" title="Permalink"></a></h3><p>The general form of the multivariate (univariate) mixed effects model with incomplete genomic data (&quot;single-step&quot; methods) for non-genotyped individuals is</p><p class="math-container">\[\mathbf{y}_{i}
=\sum_{j=1}^{p_{b}}X_{ij}\mathbf{b}_{j}+\sum_{k=1}^{p_{u}}Z_{ik}\mathbf{u}_{k}+
\sum_{l=1}^{p}\hat{M_{il}}\boldsymbol{\alpha}_{l}+\sum_{m=1}^{p_{\epsilon}}Z_{n[i,m]}\boldsymbol{\epsilon}_{m}+\boldsymbol{e}_{i} (2),\]</p><p>where <span>$\mathbf{y}_{i}$</span> is a vector of phenotypes of <span>$t$</span> traits for non-genotyped individual <span>$i$</span>;  <span>$\hat{{M}_{il}}$</span> is the imputed genotype covariate at locus <span>$l$</span> for non-genotyped individual <span>$i$</span>, <span>$Z_{n[i,m]}$</span> is the incidence matrix covariate corresponding to the <span>$m$</span>th imputation residual for individual <span>$i$</span> and <span>$\boldsymbol{\epsilon}_i$</span> is a vector of imputation residuals. <span>$W_{im}$</span> is the incidence matrix covariate corresponding to the <span>$m$</span>th random effect for individual <span>$i$</span>. That vector of imputation residuals, <span>$\boldsymbol{\epsilon}=\begin{bmatrix}\boldsymbol{\epsilon}_{1}^{T} &amp; \boldsymbol{\epsilon}_{2}^{T} &amp; \ldots &amp; \end{bmatrix}^{T}$</span>, are a priori assumed to be <span>$N\left(0,(\mathbf{A}_{nn}-\mathbf{A}_{ng}\mathbf{A}_{gg}^{-1}\mathbf{A}_{gn})\otimes\mathbf{G}_{g}\right)$</span>, where <span>$\mathbf{A}_{nn}$</span> is the partition of the numerator relationship matrix  <span>$\mathbf{A}$</span> that corresponds to non-genotyped individuals, <span>$\mathbf{A}_{ng}$</span> or its transpose <span>$\mathbf{A}_{gn}$</span> are partitions of <span>$\mathbf{A}$</span> corresponding to relationships between non-genotyped and genotyped individuals or vice versa,  <span>$\mathbf{A}_{gg}$</span> is the  partition of <span>$\mathbf{A}$</span> that corresponds to genotyped animals, and <span>$\mathbf{G}_{g}$</span> is the additive genetic covariance matrix. All the other variables are the same as in equation (1).</p><h3 id="Priors"><a class="docs-heading-anchor" href="#Priors">Priors</a><a id="Priors-1"></a><a class="docs-heading-anchor-permalink" href="#Priors" title="Permalink"></a></h3><h4 id="Priors-for-effects-other-than-markers"><a class="docs-heading-anchor" href="#Priors-for-effects-other-than-markers">Priors for effects other than markers</a><a id="Priors-for-effects-other-than-markers-1"></a><a class="docs-heading-anchor-permalink" href="#Priors-for-effects-other-than-markers" title="Permalink"></a></h4><p>The fixed effects are assigned flat priors. The vector of random effects, <span>$\mathbf{u}=\begin{bmatrix}\mathbf{u}_{1}^{T} &amp; \mathbf{u}_{2}^{T} &amp; \ldots &amp; \mathbf{u}_{p_{2}}^{T}\end{bmatrix}^{T}$</span>, are a priori assumed to be <span>$N\left(0,\mathbf{A}\otimes\mathbf{G}\right)$</span> with various options for <span>$\mathbf{A}$</span>. For example, <span>$\mathbf{A}$</span> could be an identity matrix if <span>$\boldsymbol{u}_{k}$</span> is assumed to be independently and identically distributed. <span>$\mathbf{A}$</span> can be  the numerator relationship matrix, when <span>$\boldsymbol{u}$</span> is a vector of polygenic effects and <span>$\mathbf{G}$</span> represents the additive-genetic variance not explained by molecular markers. Note that <span>$\boldsymbol{u}$</span> can also be a concatenation of vectors of different types of random effects, such as litter, pen, polygenic and maternal effects. The vector <span>$\boldsymbol{e}_{i}$</span> of residuals are a priori assumed to be independently and identically following multivariate normal distributions with null mean and covariance matrix <span>$\mathbf{R}$</span>, which in turn is a priori assumed to have an inverse Wishart prior distribution, <span>$W_{t}^{-1}\left(\mathbf{S}_{e},\nu_{e}\right)$</span>. Note that when number of traits <span>$t=1$</span>, the priors for <span>$\mathbf{G}$</span> and <span>$\mathbf{R}$</span> in single-trait analyses follow scaled inverted chi-square distributions.</p><h4 id="Priors-for-marker-effects"><a class="docs-heading-anchor" href="#Priors-for-marker-effects">Priors for marker effects</a><a id="Priors-for-marker-effects-1"></a><a class="docs-heading-anchor-permalink" href="#Priors-for-marker-effects" title="Permalink"></a></h4><h5 id="single-trait-BayesA"><a class="docs-heading-anchor" href="#single-trait-BayesA">single-trait BayesA</a><a id="single-trait-BayesA-1"></a><a class="docs-heading-anchor-permalink" href="#single-trait-BayesA" title="Permalink"></a></h5><p>The prior assumption is that marker effects have identical and independent univariate-t distributions each with a null mean, scale parameter <span>$S^2_{\alpha}$</span> and <span>$\nu$</span> degrees of freedom. This is equivalent to assuming that the marker effect at locus <span>$i$</span> has a univariate normal with null mean and unknown, locus-specific variance <span>$\sigma^2_i$</span>, which in turn is assigned a scaled inverse chi-square prior with scale parameter <span>$S^2_{\alpha}$</span> and <span>$\nu_{\alpha}$</span> degrees of freedom.</p><h5 id="single-trait-BayesB"><a class="docs-heading-anchor" href="#single-trait-BayesB">single-trait BayesB</a><a id="single-trait-BayesB-1"></a><a class="docs-heading-anchor-permalink" href="#single-trait-BayesB" title="Permalink"></a></h5><p>In BayesB, the prior assumption is that marker effects have identical and independent mixture distributions, where each has a point mass at zero with probability <span>$\pi$</span> and a univariate-t distribution with probability <span>$1-\pi$</span> having a null mean, scale parameter <span>$S^2_{\alpha}$</span> and <span>$\nu$</span> degrees of freedom. Thus, BayesA is a special case of BayesB with <span>$\pi=0$</span>. Further, as in BayesA, the t-distribution in BayesB is equivalent to a univariate normal with null mean and unknown, locus-specific variance, which in turn is assigned a scaled inverse chi-square prior with scale parameter <span>$S^2_{\alpha}$</span> and <span>$\nu_{\alpha}$</span> degrees of freedom. <em>(A fast and efficient Gibbs sampler was implemented for BayesB in JWAS.)</em></p><h5 id="single-trait-BayesC-and-BayesC\\pi"><a class="docs-heading-anchor" href="#single-trait-BayesC-and-BayesC\\pi">single-trait BayesC and BayesC<span>$\pi$</span></a><a id="single-trait-BayesC-and-BayesC\\pi-1"></a><a class="docs-heading-anchor-permalink" href="#single-trait-BayesC-and-BayesC\\pi" title="Permalink"></a></h5><p>In BayesC, the prior assumption is that marker effects have identical and independent mixture distributions, where each has a point mass at zero with probability <span>$\pi$</span> and a univariate-normal distribution with probability <span>$1-\pi$</span> having a null mean and variance <span>$\sigma^2_{\alpha}$</span>, which in turn has a scaled inverse chi-square prior with scale parameter <span>$S^2_{\alpha}$</span> and <span>$\nu_{\alpha}$</span> degrees of freedom. In addition to the above assumptions, in BayesC <span>$\pi$</span>, <span>$\pi$</span> is treated as unknown with a uniform prior.</p><h5 id="multiple-trait-Bayesian-Alphabet"><a class="docs-heading-anchor" href="#multiple-trait-Bayesian-Alphabet">multiple-trait Bayesian Alphabet</a><a id="multiple-trait-Bayesian-Alphabet-1"></a><a class="docs-heading-anchor-permalink" href="#multiple-trait-Bayesian-Alphabet" title="Permalink"></a></h5><p>In multi-trait BayesC<span>$\Pi$</span>, the prior for <span>$\alpha_{lk}$</span>, the marker effect of trait <span>$k$</span> for locus <span>$l$</span>, is a mixture with a point mass at zero and a univariate normal distribution conditional on <span>$\sigma_{k}^{2}$</span>:</p><p class="math-container">\[\alpha_{lk}\mid\pi_{k},\sigma_{k}^{2}  \begin{cases}
\sim N\left(0,\,\sigma_{k}^{2}\right)  &amp; \:probability\;(1-\pi_{k})\\
0  &amp; \:probability\;\pi_{k}
\end{cases}\]</p><p>and the covariance between effects for traits <span>$k$</span> and <span>$k&#39;$</span> at the same locus, i.e., <span>$\alpha_{lk}$</span> and <span>$\alpha_{lk^{&#39;}}$</span> is</p><p class="math-container">\[cov\left(\alpha_{lk},\alpha_{lk^{&#39;}}\mid\sigma_{kk^{&#39;}}\right)=\begin{cases}
\sigma_{kk^{&#39;}} &amp; \:if\:both\,\alpha_{lk}\neq0\:and\:\alpha_{lk^{&#39;}}\neq0\\
0 &amp; \:otherwise
\end{cases}.\]</p><p>The vector of marker effects at a particular locus <span>$\boldsymbol{\alpha}_{l}$</span> is written as <span>$\boldsymbol{\alpha}_{l}=\boldsymbol{D}_{l}\boldsymbol{\beta}_{l}$</span>, where <span>$\boldsymbol{D}_{l}$</span> is a diagonal matrix with elements <span>$diag\left(\boldsymbol{D}_{l}\right)=\boldsymbol{\delta}_{l}=\left(\delta_{l1},\delta_{l2},\delta_{l3}\ldots\delta_{lt}\right)$</span>, where <span>$\delta_{lk}$</span> is an indicator variable indicating whether the marker effect of locus <span>$l$</span> for trait <span>$k$</span> is zero or non-zero, and the vector <span>$\boldsymbol{\beta}_{l}$</span> follows a multivariate normal distribution with null mean and covariance matrix <span>$\boldsymbol{G}$</span>. The covariance matrix <span>$\boldsymbol{G}$</span> is <span>$a$</span> <span>$priori$</span> assumed to follow an inverse Wishart distribution, <span>$W_{t}^{-1}\left(\mathbf{S}_{\beta},\nu_{\beta}\right)$</span>.</p><p>In the most general case, any marker effect might be zero for any possible combination of <span>$t$</span> traits resulting in <span>$2^{t}$</span> possible combinations of <span>$\boldsymbol{\delta}_{l}$</span>. For example, in a <span>$t$</span>=2 trait model, there are <span>$2^{t}=4$</span> combinations for  <span>$\boldsymbol{\delta}_{l}$</span>: <span>$(0,\,0)$</span>, <span>$(0,\,1)$</span>, <span>$(1,\,0)$</span>, <span>$(1,\,1)$</span>. Suppose in general we use numerical labels &quot;1&quot;, &quot;2&quot;,<span>$\ldots$</span>, &quot;<span>$l$</span>&quot; for the <span>$2^{t}$</span> possible outcomes for  <span>$\boldsymbol{\delta}_{l}$</span>, then the prior for  <span>$\boldsymbol{\delta}_{l}$</span> is a categorical distribution</p><p class="math-container">\[p\left(\boldsymbol{\delta}_{l}=``i&quot;\right)= \Pi_{1}I\left(\boldsymbol{\delta}_{l}=``1&quot;\right)+\Pi_{2}I\left(\boldsymbol{\delta}_{l}=``2&quot;\right)+...+\Pi_{l}I\left(\boldsymbol{\delta}_{l}=``l&quot;\right),\]</p><p>where <span>$\sum_{i=1}^{l}\Pi_{i}=1$</span> with <span>$\Pi_{i}$</span> being the prior probability that the vector <span>$\boldsymbol{\delta}_{l}$</span> corresponds to the vector labelled <span>$&quot;i&quot;$</span>. A Dirichlet distribution with all parameters equal to one, i.e., a uniform distribution, can be used for the prior for <span>$\boldsymbol{\Pi}=\left(\Pi_{1},\Pi_{2},...,\Pi_{l}\right)$</span>.   </p><p>The differences in multi-trait BayesB method is that the prior for <span>$\boldsymbol{\beta}_{l}$</span> is a multivariate t distribution, rather than a multivariate normal distribution. This is equivalent to assuming <span>$\boldsymbol{\beta}_{l}$</span> has a multivariate normal distribution with null mean and locus-specific covariance matrix <span>$\boldsymbol{G}_{l}$</span>, which is assigned an inverse Wishart prior, <span>$W_{t}^{-1}\left(\mathbf{S}_{\beta},\nu_{\beta}\right)$</span>. Multi-trait BayesA method is a special case of multi-trait BayesB method where <span>$\boldsymbol{\delta}_{l}$</span> is always a vector of ones.</p><blockquote><h4>references</h4><ul><li>Meuwissen T, Hayes B, Goddard M. Prediction of total genetic value using genome-wide dense marker maps. Genetics, 2001,157:1819–1829.</li><li>Fernando R, Garrick D. Bayesian methods applied to GWAS. Methods Mol Biol. 2013, 1019:237–274.</li><li>Cheng H, Garrick D, Fernando R. A fast and efficient Gibbs sampler for BayesB in whole- genome analyses. Genetics Selection Evolution, 2015, 47:80.</li><li>Fernando R, Dekkers J, Garrick D. A class of Bayesian methods to combine large numbers of genotyped and non-genotyped animals for whole-genome analyses. Genetics Selection Evolution, 2015, 46(1), 50.</li><li>Fernando R, Cheng H, Golden B, Garrick D.. Computational strategies for alternative single-step Bayesian regression models with large numbers of genotyped and non-genotyped animals. Genetics Selection Evolution, 2016, 48(1), 96.</li><li>Cheng H, Kizilkaya K, Zeng J, Garrick D, Fernando R. Genomic Prediction from Multiple-trait Bayesian Regression Methods using Mixture Priors. Genetics. 2018, 209(1).</li></ul></blockquote></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">« Home</a><a class="docs-footer-nextpage" href="../../citing/citing/">Citing »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 7 October 2025 22:19">Tuesday 7 October 2025</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
