{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<ul class=\"breadcrumb\">\n",
    "  <li><a href=\"1_Conventional_Linear_Mixed_Model.ipynb\">Bayesian Linear Mixed Models (Conventional)</a></li>\n",
    "  <li><a href=\"2_Linear_Additive_Genetic_Model.ipynb\">Bayesian Linear Additive Genetic Model</a></li> \n",
    "  <li><a href=\"3_Genomic_Linear_Mixed_Model.ipynb\">Bayesian Linear Mixed Models (Genomic Data)</a></li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<button type=\"button\" class=\"btn btn-lg btn-primary\">Step 1: Load Packages</button> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "using JWAS,JWAS.Datasets,DataFrames,CSV"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<button type=\"button\" class=\"btn btn-lg btn-primary\">Step 2: Read data</button> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mThe delimiter in pedigree.txt is ','.\u001b[39m\n",
      "Finished!\n"
     ]
    }
   ],
   "source": [
    "phenofile  = Datasets.dataset(\"example\",\"phenotypes.txt\")\n",
    "pedfile    = Datasets.dataset(\"example\",\"pedigree.txt\")\n",
    "genofile   = Datasets.dataset(\"example\",\"genotypes.txt\")\n",
    "\n",
    "phenotypes = CSV.read(phenofile,delim = ',',header=true,missingstrings=[\"NA\"])\n",
    "pedigree   = get_pedigree(pedfile,separator=\",\",header=true);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>ID</th><th>y1</th><th>y2</th><th>y3</th><th>x1</th><th>x2</th><th>x3</th><th>dam</th></tr><tr><th></th><th>String</th><th>Float64</th><th>Float64⍰</th><th>Float64⍰</th><th>Float64</th><th>Int64</th><th>String</th><th>String</th></tr></thead><tbody><p>5 rows × 8 columns</p><tr><th>1</th><td>a1</td><td>-0.06</td><td>3.58</td><td>-1.18</td><td>0.9</td><td>2</td><td>m</td><td>0</td></tr><tr><th>2</th><td>a3</td><td>-2.07</td><td>3.19</td><td>missing</td><td>0.7</td><td>2</td><td>f</td><td>0</td></tr><tr><th>3</th><td>a4</td><td>-2.63</td><td>6.97</td><td>-0.83</td><td>0.6</td><td>1</td><td>m</td><td>a2</td></tr><tr><th>4</th><td>a5</td><td>2.31</td><td>missing</td><td>-1.52</td><td>0.4</td><td>2</td><td>m</td><td>a2</td></tr><tr><th>5</th><td>a6</td><td>0.93</td><td>4.87</td><td>-0.01</td><td>5.0</td><td>2</td><td>f</td><td>a3</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccccccc}\n",
       "\t& ID & y1 & y2 & y3 & x1 & x2 & x3 & dam\\\\\n",
       "\t\\hline\n",
       "\t& String & Float64 & Float64⍰ & Float64⍰ & Float64 & Int64 & String & String\\\\\n",
       "\t\\hline\n",
       "\t1 & a1 & -0.06 & 3.58 & -1.18 & 0.9 & 2 & m & 0 \\\\\n",
       "\t2 & a3 & -2.07 & 3.19 &  & 0.7 & 2 & f & 0 \\\\\n",
       "\t3 & a4 & -2.63 & 6.97 & -0.83 & 0.6 & 1 & m & a2 \\\\\n",
       "\t4 & a5 & 2.31 &  & -1.52 & 0.4 & 2 & m & a2 \\\\\n",
       "\t5 & a6 & 0.93 & 4.87 & -0.01 & 5.0 & 2 & f & a3 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "5×8 DataFrame. Omitted printing of 1 columns\n",
       "│ Row │ ID     │ y1      │ y2       │ y3       │ x1      │ x2    │ x3     │\n",
       "│     │ \u001b[90mString\u001b[39m │ \u001b[90mFloat64\u001b[39m │ \u001b[90mFloat64⍰\u001b[39m │ \u001b[90mFloat64⍰\u001b[39m │ \u001b[90mFloat64\u001b[39m │ \u001b[90mInt64\u001b[39m │ \u001b[90mString\u001b[39m │\n",
       "├─────┼────────┼─────────┼──────────┼──────────┼─────────┼───────┼────────┤\n",
       "│ 1   │ a1     │ -0.06   │ 3.58     │ -1.18    │ 0.9     │ 2     │ m      │\n",
       "│ 2   │ a3     │ -2.07   │ 3.19     │ \u001b[90mmissing\u001b[39m  │ 0.7     │ 2     │ f      │\n",
       "│ 3   │ a4     │ -2.63   │ 6.97     │ -0.83    │ 0.6     │ 1     │ m      │\n",
       "│ 4   │ a5     │ 2.31    │ \u001b[90mmissing\u001b[39m  │ -1.52    │ 0.4     │ 2     │ m      │\n",
       "│ 5   │ a6     │ 0.93    │ 4.87     │ -0.01    │ 5.0     │ 2     │ f      │"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first(phenotypes,5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"span5 alert alert-success\">\n",
    " <font size=\"5\" face=\"Georgia\">Univariate Linear Mixed Model (Genomic data)</font> \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<button type=\"button\" class=\"btn btn-lg btn-primary\">Step 3: Build Model Equations</button> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_equation1  =\"y1 = intercept + x1*x3 + x2 + x3 + ID + dam\";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "R      = 1.0\n",
    "model1 = build_model(model_equation1,R);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<button type=\"button\" class=\"btn btn-lg btn-primary\">Step 4: Set Factors or Covariates</button> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_covariate(model1,\"x1\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<button type=\"button\" class=\"btn btn-lg btn-primary\">Step 5: Set Random or Fixed Effects</button> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "G1 = 1.0\n",
    "G2 = [1.0 0.5\n",
    "      0.5 1.0]\n",
    "set_random(model1,\"x2\",G1);\n",
    "set_random(model1,\"ID dam\",pedigree,G2);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<button type=\"button\" class=\"btn btn-lg btn-primary\">Step 6: Use Genomic Information</button> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mThe delimiter in genotypes.txt is ','.\u001b[39m\n",
      "\u001b[31mThe header (marker IDs) is provided in genotypes.txt.\u001b[39m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: readtable is deprecated, use CSV.read from the CSV package instead\n",
      "│   caller = ip:0x0\n",
      "└ @ Core :-1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5 markers on 7 individuals were added.\n"
     ]
    }
   ],
   "source": [
    "G3 =1.0\n",
    "add_genotypes(model1,genofile,G3,separator=',');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<button type=\"button\" class=\"btn btn-lg btn-primary\">Step 7: Run Analysis</button> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The prior for marker effects variance is calculated from the genetic variance and π.\n",
      "The mean of the prior for the marker effects variance is: 0.492462\n",
      "\n",
      "\n",
      "\u001b[0m\u001b[1mA Linear Mixed Model was build using model equations:\u001b[22m\n",
      "\n",
      "y1 = intercept + x1*x3 + x2 + x3 + ID + dam\n",
      "\n",
      "\u001b[0m\u001b[1mModel Information:\u001b[22m\n",
      "\n",
      "Term            C/F          F/R            nLevels\n",
      "intercept       factor       fixed                1\n",
      "x1*x3           interaction  fixed                2\n",
      "x2              factor       random               2\n",
      "x3              factor       fixed                2\n",
      "ID              factor       random              12\n",
      "dam             factor       random              12\n",
      "\n",
      "\u001b[0m\u001b[1mMCMC Information:\u001b[22m\n",
      "\n",
      "methods                                      BayesC\n",
      "chain_length                                   5000\n",
      "burnin                                            0\n",
      "estimatePi                                     true\n",
      "estimateScale                                 false\n",
      "starting_value                                false\n",
      "printout_frequency                             5001\n",
      "output_samples_frequency                        100\n",
      "constraint                                    false\n",
      "missing_phenotypes                             true\n",
      "update_priors_frequency                           0\n",
      "\n",
      "\u001b[0m\u001b[1mHyper-parameters Information:\u001b[22m\n",
      "\n",
      "random effect variances (x2):                 [1.0]\n",
      "residual variances:                           1.000\n",
      "genetic variances (polygenic):\n",
      "                                 [1.0 0.5; 0.5 1.0]\n",
      "genetic variances (genomic):                  1.000\n",
      "marker effect variances:                      0.492\n",
      "π                                               0.0\n",
      "\n",
      "\u001b[0m\u001b[1mDegree of freedom for hyper-parameters:\u001b[22m\n",
      "\n",
      "residual variances:                           4.000\n",
      "iid random effect variances:                  4.000\n",
      "polygenic effect variances:                   4.000\n",
      "marker effect variances:                      4.000\n",
      "\n",
      "\n",
      "\n",
      "\u001b[32mThe file MCMC_samples_residual_variance.txt is created to save MCMC samples for residual_variance.\u001b[39m\n",
      "\u001b[32mThe file MCMC_samples_polygenic_effects_variance.txt is created to save MCMC samples for polygenic_effects_variance.\u001b[39m\n",
      "\u001b[32mThe file MCMC_samples_marker_effects_y1.txt is created to save MCMC samples for marker_effects_y1.\u001b[39m\n",
      "\u001b[32mThe file MCMC_samples_marker_effects_variances.txt is created to save MCMC samples for marker_effects_variances.\u001b[39m\n",
      "\u001b[32mThe file MCMC_samples_pi.txt is created to save MCMC samples for pi.\u001b[39m\n",
      "\u001b[32mThe file MCMC_samples_1:x2.txt is created to save MCMC samples for 1:x2.\u001b[39m\n",
      "\u001b[32mThe file MCMC_samples_x2_variances.txt is created to save MCMC samples for x2_variances.\u001b[39m\n",
      "\u001b[32mThe file MCMC_samples_EBV_y1.txt is created to save MCMC samples for EBV_y1.\u001b[39m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mrunning MCMC for BayesC...100%|█████████████████████████| Time: 0:00:01\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "outputMCMCsamples(model1,\"x2\")\n",
    "out1=runMCMC(model1,phenotypes,methods=\"BayesC\",estimatePi=true,chain_length=5000,output_samples_frequency=100);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "search: \u001b[0m\u001b[1mG\u001b[22m\u001b[0m\u001b[1mW\u001b[22m\u001b[0m\u001b[1mA\u001b[22m\u001b[0m\u001b[1mS\u001b[22m\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/latex": [
       "\\begin{verbatim}\n",
       "GWAS(marker_effects_file;header=false)\n",
       "\\end{verbatim}\n",
       "Compute the model frequency for each marker (the probability the marker is included in the model) using samples of marker effects stored in \\textbf{marker\\emph{effects}file}.\n",
       "\n",
       "\\rule{\\textwidth}{1pt}\n",
       "\\begin{verbatim}\n",
       "GWAS(marker_effects_file,model;header=true,window_size=\"1 Mb\",threshold=0.001)\n",
       "\\end{verbatim}\n",
       "run genomic window-based GWAS without marker locations\n",
       "\n",
       "\\begin{itemize}\n",
       "\\item MCMC samples of marker effects are stored in \\textbf{marker\\emph{effects}file} with delimiter ','.\n",
       "\n",
       "\n",
       "\\item \\textbf{window\\_size} is either a constant (identical number of markers in each window) or an array of number of markers in each window\n",
       "\n",
       "\n",
       "\\item \\textbf{model} is either the model::MME used in analysis or the genotypic covariate matrix M::Array\n",
       "\n",
       "\n",
       "\\item File format:\n",
       "\n",
       "\\end{itemize}\n",
       "\\rule{\\textwidth}{1pt}\n",
       "\\begin{verbatim}\n",
       "GWAS(marker_effects_file,map_file,model;header=true,window_size=\"1 Mb\",threshold=0.001)\n",
       "\\end{verbatim}\n",
       "run genomic window-based GWAS\n",
       "\n",
       "\\begin{itemize}\n",
       "\\item MCMC samples of marker effects are stored in \\textbf{marker\\emph{effects}file} with delimiter ','.\n",
       "\n",
       "\n",
       "\\item \\textbf{model} is either the model::MME used in analysis or the genotypic cavariate matrix M::Array\n",
       "\n",
       "\n",
       "\\item \\textbf{map\\_file} has the (sorted) marker position information with delimiter ','.\n",
       "\n",
       "\n",
       "\\item File format:\n",
       "\n",
       "\\end{itemize}\n",
       "\\begin{verbatim}\n",
       "markerID,chromosome,position\n",
       "m1,1,16977\n",
       "m2,1,434311\n",
       "m3,1,1025513\n",
       "m4,2,70350\n",
       "m5,2,101135\n",
       "\\end{verbatim}\n"
      ],
      "text/markdown": [
       "```\n",
       "GWAS(marker_effects_file;header=false)\n",
       "```\n",
       "\n",
       "Compute the model frequency for each marker (the probability the marker is included in the model) using samples of marker effects stored in **marker*effects*file**.\n",
       "\n",
       "---\n",
       "\n",
       "```\n",
       "GWAS(marker_effects_file,model;header=true,window_size=\"1 Mb\",threshold=0.001)\n",
       "```\n",
       "\n",
       "run genomic window-based GWAS without marker locations\n",
       "\n",
       "  * MCMC samples of marker effects are stored in **marker*effects*file** with delimiter ','.\n",
       "  * **window_size** is either a constant (identical number of markers in each window) or an array of number of markers in each window\n",
       "  * **model** is either the model::MME used in analysis or the genotypic covariate matrix M::Array\n",
       "  * File format:\n",
       "\n",
       "---\n",
       "\n",
       "```\n",
       "GWAS(marker_effects_file,map_file,model;header=true,window_size=\"1 Mb\",threshold=0.001)\n",
       "```\n",
       "\n",
       "run genomic window-based GWAS\n",
       "\n",
       "  * MCMC samples of marker effects are stored in **marker*effects*file** with delimiter ','.\n",
       "  * **model** is either the model::MME used in analysis or the genotypic cavariate matrix M::Array\n",
       "  * **map_file** has the (sorted) marker position information with delimiter ','.\n",
       "  * File format:\n",
       "\n",
       "```\n",
       "markerID,chromosome,position\n",
       "m1,1,16977\n",
       "m2,1,434311\n",
       "m3,1,1025513\n",
       "m4,2,70350\n",
       "m5,2,101135\n",
       "```\n"
      ],
      "text/plain": [
       "\u001b[36m  GWAS(marker_effects_file;header=false)\u001b[39m\n",
       "\n",
       "  Compute the model frequency for each marker (the probability the marker is\n",
       "  included in the model) using samples of marker effects stored in\n",
       "  \u001b[1mmarker\u001b[4meffects\u001b[24mfile\u001b[22m.\n",
       "\n",
       "  ────────────────────────────────────────────────────────────────────────────\n",
       "\n",
       "\u001b[36m  GWAS(marker_effects_file,model;header=true,window_size=\"1 Mb\",threshold=0.001)\u001b[39m\n",
       "\n",
       "  run genomic window-based GWAS without marker locations\n",
       "\n",
       "    •    MCMC samples of marker effects are stored in \u001b[1mmarker\u001b[4meffects\u001b[24mfile\u001b[22m\n",
       "        with delimiter ','.\n",
       "\n",
       "    •    \u001b[1mwindow_size\u001b[22m is either a constant (identical number of markers in\n",
       "        each window) or an array of number of markers in each window\n",
       "\n",
       "    •    \u001b[1mmodel\u001b[22m is either the model::MME used in analysis or the genotypic\n",
       "        covariate matrix M::Array\n",
       "\n",
       "    •    File format:\n",
       "\n",
       "  ────────────────────────────────────────────────────────────────────────────\n",
       "\n",
       "\u001b[36m  GWAS(marker_effects_file,map_file,model;header=true,window_size=\"1 Mb\",threshold=0.001)\u001b[39m\n",
       "\n",
       "  run genomic window-based GWAS\n",
       "\n",
       "    •    MCMC samples of marker effects are stored in \u001b[1mmarker\u001b[4meffects\u001b[24mfile\u001b[22m\n",
       "        with delimiter ','.\n",
       "\n",
       "    •    \u001b[1mmodel\u001b[22m is either the model::MME used in analysis or the genotypic\n",
       "        cavariate matrix M::Array\n",
       "\n",
       "    •    \u001b[1mmap_file\u001b[22m has the (sorted) marker position information with\n",
       "        delimiter ','.\n",
       "\n",
       "    •    File format:\n",
       "\n",
       "\u001b[36m  markerID,chromosome,position\u001b[39m\n",
       "\u001b[36m  m1,1,16977\u001b[39m\n",
       "\u001b[36m  m2,1,434311\u001b[39m\n",
       "\u001b[36m  m3,1,1025513\u001b[39m\n",
       "\u001b[36m  m4,2,70350\u001b[39m\n",
       "\u001b[36m  m5,2,101135\u001b[39m"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "?GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"MCMC_samples_marker_effects_y1.txt\""
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marker_effects_file=\"MCMC_samples_marker_effects_y1.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5×2 Array{Any,2}:\n",
       " \"m1\"  0.62\n",
       " \"m2\"  0.62\n",
       " \"m3\"  0.6 \n",
       " \"m4\"  0.7 \n",
       " \"m5\"  0.6 "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GWAS(marker_effects_file,header=true) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>window</th><th>chr</th><th>wStart</th><th>wEnd</th><th>start_SNP</th><th>end_SNP</th><th>numSNP</th><th>prGenVar</th><th>WPPA</th></tr><tr><th></th><th>Int64</th><th>String</th><th>Int64</th><th>Int64</th><th>Int64</th><th>Int64</th><th>Int64</th><th>Float64</th><th>Float64</th></tr></thead><tbody><p>3 rows × 10 columns (omitted printing of 1 columns)</p><tr><th>1</th><td>1</td><td>1</td><td>0</td><td>1000000</td><td>16977</td><td>434311</td><td>2</td><td>42.36</td><td>0.82</td></tr><tr><th>2</th><td>3</td><td>2</td><td>0</td><td>1000000</td><td>70350</td><td>101135</td><td>2</td><td>36.57</td><td>0.82</td></tr><tr><th>3</th><td>2</td><td>1</td><td>1000000</td><td>2000000</td><td>1025513</td><td>1025513</td><td>1</td><td>26.98</td><td>0.58</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccccccccc}\n",
       "\t& window & chr & wStart & wEnd & start\\_SNP & end\\_SNP & numSNP & prGenVar & WPPA & \\\\\n",
       "\t\\hline\n",
       "\t& Int64 & String & Int64 & Int64 & Int64 & Int64 & Int64 & Float64 & Float64 & \\\\\n",
       "\t\\hline\n",
       "\t1 & 1 & 1 & 0 & 1000000 & 16977 & 434311 & 2 & 42.36 & 0.82 & $\\dots$ \\\\\n",
       "\t2 & 3 & 2 & 0 & 1000000 & 70350 & 101135 & 2 & 36.57 & 0.82 & $\\dots$ \\\\\n",
       "\t3 & 2 & 1 & 1000000 & 2000000 & 1025513 & 1025513 & 1 & 26.98 & 0.58 & $\\dots$ \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "3×10 DataFrame. Omitted printing of 3 columns\n",
       "│ Row │ window │ chr    │ wStart  │ wEnd    │ start_SNP │ end_SNP │ numSNP │\n",
       "│     │ \u001b[90mInt64\u001b[39m  │ \u001b[90mString\u001b[39m │ \u001b[90mInt64\u001b[39m   │ \u001b[90mInt64\u001b[39m   │ \u001b[90mInt64\u001b[39m     │ \u001b[90mInt64\u001b[39m   │ \u001b[90mInt64\u001b[39m  │\n",
       "├─────┼────────┼────────┼─────────┼─────────┼───────────┼─────────┼────────┤\n",
       "│ 1   │ 1      │ 1      │ 0       │ 1000000 │ 16977     │ 434311  │ 2      │\n",
       "│ 2   │ 3      │ 2      │ 0       │ 1000000 │ 70350     │ 101135  │ 2      │\n",
       "│ 3   │ 2      │ 1      │ 1000000 │ 2000000 │ 1025513   │ 1025513 │ 1      │"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_file = Datasets.dataset(\"example\",\"map.txt\")\n",
    "GWAS(marker_effects_file,map_file,model1,header=true,window_size=\"1 Mb\",threshold=0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Julia 1.1.0",
   "language": "julia",
   "name": "julia-1.1"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
