<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Citing · JWAS.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="JWAS.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">JWAS.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../theory/theory/">Some Theory</a></li><li class="is-active"><a class="tocitem" href>Citing</a><ul class="internal"><li><a class="tocitem" href="#Software"><span>Software</span></a></li><li><a class="tocitem" href="#Methods"><span>Methods</span></a></li></ul></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../manual/getstarted/">Get Started</a></li><li><a class="tocitem" href="../../manual/workflow/">Workflow</a></li><li><a class="tocitem" href="../../manual/public/">Public</a></li><li><a class="tocitem" href="../../manual/internals/">Internals</a></li></ul></li><li><span class="tocitem">Mixed Effects Neural Networks (NNMM)</span><ul><li><a class="tocitem" href="../../nnmm/Part1_introduction/">Part 1. introduction</a></li><li><a class="tocitem" href="../../nnmm/Part2_NNMM/">Part 2. NNMM</a></li><li><a class="tocitem" href="../../nnmm/Part3_NNMMwithIntermediateOmics/">Part 3. NNMM with intermediate omics</a></li><li><a class="tocitem" href="../../nnmm/Part4_PartialConnectedNeuralNetwork/">Part 4. partial connected neural network</a></li><li><a class="tocitem" href="../../nnmm/Part5_UserDefinedNonlinearFunction/">Part 5. user-defined onlinear function</a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/examples/">Examples</a></li></ul></li><li><a class="tocitem" href="../../FrequentlyAskedQuestions/FrequentlyAskedQuestions/">Frequently Asked Questions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Citing</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Citing</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/reworkhow/JWAS.jl/blob/master/docs/src/citing/citing.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Citing"><a class="docs-heading-anchor" href="#Citing">Citing</a><a id="Citing-1"></a><a class="docs-heading-anchor-permalink" href="#Citing" title="Permalink"></a></h1><h2 id="Software"><a class="docs-heading-anchor" href="#Software">Software</a><a id="Software-1"></a><a class="docs-heading-anchor-permalink" href="#Software" title="Permalink"></a></h2><ul><li>Cheng, H., Fernando, R. L., and Garrick, D. J. 2018 JWAS: Julia implementation of whole-genome analysis software. Proceedings of the World Congress on Genetics Applied to Livestock Production,11.859. Auckland, New Zealand.</li></ul><h2 id="Methods"><a class="docs-heading-anchor" href="#Methods">Methods</a><a id="Methods-1"></a><a class="docs-heading-anchor-permalink" href="#Methods" title="Permalink"></a></h2><h3 id="single-trait-BayesA"><a class="docs-heading-anchor" href="#single-trait-BayesA">single-trait BayesA</a><a id="single-trait-BayesA-1"></a><a class="docs-heading-anchor-permalink" href="#single-trait-BayesA" title="Permalink"></a></h3><ul><li>Meuwissen T, Hayes B, Goddard M. 2001 Prediction of total genetic value using genome-wide dense marker maps. Genetics, 157:1819–1829.</li></ul><h3 id="single-trait-BayesB"><a class="docs-heading-anchor" href="#single-trait-BayesB">single-trait BayesB</a><a id="single-trait-BayesB-1"></a><a class="docs-heading-anchor-permalink" href="#single-trait-BayesB" title="Permalink"></a></h3><ul><li>Meuwissen T, Hayes B, Goddard M. 2001 Prediction of total genetic value using genome-wide dense marker maps. Genetics, 157:1819–1829.</li><li>Cheng H, Garrick D, Fernando R. 2015 A fast and efficient Gibbs sampler for BayesB in whole-genome analyses. Genet Sel Evol, 47:80.</li></ul><h3 id="single-trait-BayesC"><a class="docs-heading-anchor" href="#single-trait-BayesC">single-trait BayesC</a><a id="single-trait-BayesC-1"></a><a class="docs-heading-anchor-permalink" href="#single-trait-BayesC" title="Permalink"></a></h3><ul><li>Habier D, Fernando R, Kizilkaya K, Garrick D. 2011 Extension of the bayesian alphabet for genomic selection. BMC Bioinformatics, 12(1), 186.</li></ul><h3 id="single-trait-Bayesian-LASSO"><a class="docs-heading-anchor" href="#single-trait-Bayesian-LASSO">single-trait Bayesian LASSO</a><a id="single-trait-Bayesian-LASSO-1"></a><a class="docs-heading-anchor-permalink" href="#single-trait-Bayesian-LASSO" title="Permalink"></a></h3><ul><li>Gianola, D., &amp; Fernando, R. L. (2019). A Multiple-Trait Bayesian Lasso for Genome-Enabled Analysis and Prediction of Complex Traits. Genetics, 214(2), genetics.302934.2019–331.</li></ul><h3 id="multi-trait-Bayesian-Alphabet"><a class="docs-heading-anchor" href="#multi-trait-Bayesian-Alphabet">multi-trait Bayesian Alphabet</a><a id="multi-trait-Bayesian-Alphabet-1"></a><a class="docs-heading-anchor-permalink" href="#multi-trait-Bayesian-Alphabet" title="Permalink"></a></h3><ul><li>Cheng H, Kizilkaya K, Zeng J, Garrick D, Fernando R 2018 Genomic Prediction from Multiple-Trait Bayesian Regression Methods Using Mixture Priors. Genetics, 209(1): 89-103.</li></ul><h3 id="single-step-Bayesian-Alphabet"><a class="docs-heading-anchor" href="#single-step-Bayesian-Alphabet">single-step Bayesian Alphabet</a><a id="single-step-Bayesian-Alphabet-1"></a><a class="docs-heading-anchor-permalink" href="#single-step-Bayesian-Alphabet" title="Permalink"></a></h3><ul><li><p>Fernando R, Dekkers J,Garrick D. 2015 A class of Bayesian methods to combine large numbers of genotyped and non-genotyped animals for whole-genome analyses. Genetics Selection Evolution, 46(1), 50.</p></li><li><p>Fernando, R. L., Cheng, H., Golden, B. L., &amp; Garrick, D. J. (2016). Computational strategies for alternative single-step Bayesian regression models with large numbers of genotyped and non-genotyped animals. Genetics Selection Evolution, 48(1), 96.</p></li></ul><h3 id="GWAS"><a class="docs-heading-anchor" href="#GWAS">GWAS</a><a id="GWAS-1"></a><a class="docs-heading-anchor-permalink" href="#GWAS" title="Permalink"></a></h3><ul><li>Fernando R, Garrick D. 2013 Bayesian methods applied to GWAS. Methods Mol Biol., 1019:237–274.</li><li>Fernando, R., Toosi, A., Wolc, A., Garrick, D., &amp; Dekkers, J. (2017). Application of Whole-Genome Prediction Methods for Genome-Wide Association Studies: A Bayesian Approach. Journal of Agricultural, Biological and Environmental Statistics, 57(4), 1–22.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../theory/theory/">« Some Theory</a><a class="docs-footer-nextpage" href="../../manual/getstarted/">Get Started »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 8 March 2024 20:20">Friday 8 March 2024</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
